base/random.jl

```diff
@@ -1154,20 +1156,55 @@ function randexp_unlikely(rng, idx, x)
     end
 end
 
-function randexp!(rng::AbstractRNG, A::Array{Float64})
-    for i in eachindex(A)
-        @inbounds A[i] = randexp(rng)
-    end
-    A
-end
+"""
+    randn!([rng], A::AbstractArray) -> A
 
-randexp!(A::Array{Float64}) = randexp!(GLOBAL_RNG, A)
-randexp(dims::Dims) = randexp!(Array(Float64, dims))
-randexp(dims::Int...) = randexp!(Array(Float64, dims))
-randexp(rng::AbstractRNG, dims::Dims) = randexp!(rng, Array(Float64, dims))
-randexp(rng::AbstractRNG, dims::Int...) = randexp!(rng, Array(Float64, dims))
+Fill the array `A` with normally-distributed (mean 0, standard deviation 1) random numbers.
+Also see the `rand` function.
+"""
+function randn! end
+
+"""
+    randexp!([rng], A::AbstractArray) -> A
+
+Fill the array `A` with random numbers following the exponential distribution (with scale 1).
+"""
+function randexp! end
+
+let Floats = Union{Float16,Float32,Float64}
+    for randfun in [:randn, :randexp]
+        randfun! = Symbol(randfun, :!)
+        @eval begin
+
+            # scalars
 
+            $randfun{T<:$Floats}(rng::AbstractRNG, ::Type{T}) = convert(T, $randfun(rng))
+            $randfun{T<:$Floats}(::Type{T}) = $randfun(GLOBAL_RNG, T)
 
+            # filling arrays
+
+            function $randfun!{T}(rng::AbstractRNG, A::AbstractArray{T})
```

Comment:
It is used below as a parameter of `$randfun`... Shall I change it to `eltype(A)`?